### PR TITLE
Fix publishing to fifo sns->sqs subscription

### LINKED ISSUE #4737: Publishing to FIFO sns topic with FIFO sqs queue subscription doesn't work
Creating a FIFO sns topic and subscribing a FIFO sqs queue works, but there's no way to publish to it since the MessageGroupId parameter is not forwarded.

Example test case:

```python
from moto import mock_sns, mock_sqs
import boto3

def test_sns_sqs():
    with mock_sns(), mock_sqs():
        sns = boto3.resource('sns')
        sqs = boto3.resource('sqs')
        queue = sqs.create_queue(
            QueueName="queue.fifo",
            Attributes={
                'FifoQueue': 'true',
                'ContentBasedDeduplication': 'true',
            },
        )

        topic = sns.create_topic(
            Name="topic.fifo",
            Attributes={
                'FifoTopic': 'true',
                'ContentBasedDeduplication': 'true',
            },
        )
        topic.subscribe(
            Protocol='sqs',
            Endpoint=queue.attributes['QueueArn'],
        )

        topic.publish(Message="message", MessageGroupId="message_group_id")

```

Traceback:

```
______________________________________________ test_sns_sqs ______________________________________________

    def test_sns_sqs():
    
        with mock_sns(), mock_sqs():
            sns = boto3.resource('sns')
            sqs = boto3.resource('sqs')
            queue = sqs.create_queue(
                QueueName="queue.fifo",
                Attributes={
                    'FifoQueue': 'true',
                    'ContentBasedDeduplication': 'true',
                },
            )
    
            topic = sns.create_topic(
                Name="topic.fifo",
                Attributes={
                    'FifoTopic': 'true',
                    'ContentBasedDeduplication': 'true',
                },
            )
            topic.subscribe(
                Protocol='sqs',
                Endpoint=queue.attributes['QueueArn'],
            )
    
>           topic.publish(Message="message", MessageGroupId="message_group_id")

app/test_main.py:71: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../.direnv/python-3.10.1/lib/python3.10/site-packages/boto3/resources/factory.py:520: in do_action
    response = action(self, *args, **kwargs)
../../.direnv/python-3.10.1/lib/python3.10/site-packages/boto3/resources/action.py:83: in __call__
    response = getattr(parent.meta.client, operation_name)(*args, **params)
../../.direnv/python-3.10.1/lib/python3.10/site-packages/botocore/client.py:491: in _api_call
    return self._make_api_call(operation_name, kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <botocore.client.SNS object at 0x7f5b1b134610>, operation_name = 'Publish'
api_params = {'Message': 'message', 'MessageGroupId': 'message_group_id', 'TopicArn': 'arn:aws:sns:eu-west-1:123456789012:topic.fifo'}

    def _make_api_call(self, operation_name, api_params):
        operation_model = self._service_model.operation_model(operation_name)
        service_name = self._service_model.service_name
        history_recorder.record(
            'API_CALL',
            {
                'service': service_name,
                'operation': operation_name,
                'params': api_params,
            },
        )
        if operation_model.deprecated:
            logger.debug(
                'Warning: %s.%s() is deprecated', service_name, operation_name
            )
        request_context = {
            'client_region': self.meta.region_name,
            'client_config': self.meta.config,
            'has_streaming_input': operation_model.has_streaming_input,
            'auth_type': operation_model.auth_type,
        }
        request_dict = self._convert_to_request_dict(
            api_params, operation_model, context=request_context
        )
    
        service_id = self._service_model.service_id.hyphenize()
        handler, event_response = self.meta.events.emit_until_response(
            'before-call.{service_id}.{operation_name}'.format(
                service_id=service_id, operation_name=operation_name
            ),
            model=operation_model,
            params=request_dict,
            request_signer=self._request_signer,
            context=request_context,
        )
    
        if event_response is not None:
            http, parsed_response = event_response
        else:
            http, parsed_response = self._make_request(
                operation_model, request_dict, request_context
            )
    
        self.meta.events.emit(
            'after-call.{service_id}.{operation_name}'.format(
                service_id=service_id, operation_name=operation_name
            ),
            http_response=http,
            parsed=parsed_response,
            model=operation_model,
            context=request_context,
        )
    
        if http.status_code >= 300:
            error_code = parsed_response.get("Error", {}).get("Code")
            error_class = self.exceptions.from_code(error_code)
>           raise error_class(parsed_response, operation_name)
E           botocore.exceptions.ClientError: An error occurred (MissingParameter) when calling the Publish operation: The request must contain the parameter MessageGroupId.

../../.direnv/python-3.10.1/lib/python3.10/site-packages/botocore/client.py:897: ClientError

```

### PR DESCRIPTION
**Pull request by sindrig**
This closes #4737

### CODE REVIEW

**Comment by sindrig** [comment]
Accidentally left an unused import in there. Removed in [b309487ed484970c01c2a6c355ddf8da22ca2274](https://github.com/spulec/moto/pull/4738/commits/b309487ed484970c01c2a6c355ddf8da22ca2274) and will need approval again

File: tests/test_sns/test_publishing_boto3.py [hunk 1]
```diff
@@ -431,6 +452,34 @@ def test_publish_message_too_long():
...
+
+
+@mock_sns
+@mock_sqs
+def test_publish_group_id_to_non_fifo():
+    sns = boto3.resource("sns", region_name="us-east-1")
+    topic = sns.create_topic(Name="topic")
+
+    with pytest.raises(ClientError):
+        topic.publish(Message="message", MessageGroupId="message_group_id")
```
Comment (bblommers) [inline tests/test_sns/test_publishing_boto3.py:477]: Same here

**Review by bblommers — CHANGES_REQUESTED** [review]

File: tests/test_sns/test_publishing_boto3.py [hunk 2]
```diff
@@ -431,6 +452,34 @@ def test_publish_message_too_long():
...
 
+@mock_sns
+def test_publish_fifo_needs_group_id():
+    sns = boto3.resource("sns", region_name="us-east-1")
+    topic = sns.create_topic(
+        Name="topic.fifo",
+        Attributes={"FifoTopic": "true", "ContentBasedDeduplication": "true",},
+    )
+
+    with pytest.raises(ClientError):
```
[the remark this answers is not in the document]
Comment (sindrig) [reply tests/test_sns/test_publishing_boto3.py]: Yep, added in https://github.com/spulec/moto/pull/4738/commits/e7859eb958dcee4f5be082dd5f7ab4fb4706b6a0

File: tests/test_sns/test_publishing_boto3.py (hunk 1 above)
Comment (sindrig) [reply tests/test_sns/test_publishing_boto3.py:482]: Ditto

File: tests/test_sns/test_publishing_boto3.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sindrig) [reply tests/test_sns/test_publishing_boto3.py]: LMK if you prefer introspecting the thrown exception as opposed to `pytest.raises(match=...)`

[the remark this answers is not in the document]
Comment (bblommers) [reply tests/test_sns/test_publishing_boto3.py]: No, this is fine - as long as we know the correct error was raised. Thanks!

**Comment by bblommers** [comment]
Looks like some other tests are still failing with these changes

**Comment by sindrig** [comment]
Should be good now :pray:

**Review by bblommers — APPROVED** [review]
LGTM! Thanks for the PR and fixing this @sindrig

**Pull request merged**